sdk/dotnet/Pulumi/Core/Output.cs

```diff
@@ -0,0 +1,219 @@
+// Copyright 2016-2018, Pulumi Corporation
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Pulumi
+{
+    public static class Output
+    {
+        public static Output<T> Create<T>([MaybeNull]T value)
+            => Create(Task.FromResult(value));
+
+        public static Output<T> Create<T>(Task<T> value)
+            => Output<T>.Create(value);
+
+        public static Output<ImmutableArray<T>> All<T>(params Input<T>[] inputs)
+            => All(ImmutableArray.CreateRange(inputs));
+
+        public static Output<ImmutableArray<T>> All<T>(ImmutableArray<Input<T>> inputs)
+            => Output<T>.All(inputs);
+
+        public static Output<(X, Y)> Tuple<X, Y>(Input<X> item1, Input<Y> item2)
+            => Tuple<X, Y, int>(item1, item2, 0).Apply(v => (v.Item1, v.Item2));
+
+        public static Output<(X, Y, Z)> Tuple<X, Y, Z>(Input<X> item1, Input<Y> item2, Input<Z> item3)

```

Comment:
Then there are ambiguities and odd deviations (because tuples/arrays are not uniform in C# like they are in TS).  i.e. in some case you'll call `.All` and will get an array. in some cases you'll get a tuple.  and that can just change unexpectedly.